File: fszmq.props/fszmq.props.cs/Generators.cs
namespace fszmq.props
{
  using FsCheck;
  using System;
  using System.Text;
  using System.Collections;
  using System.Collections.Generic;
  using System.Linq;
  
  using GenMessage    = FsCheck.Gen<fszmq.Message>;
  using GenVersion    = FsCheck.Gen<fszmq.Version>;

  /// <summary>
  /// A byte[] whose length is evenly divisible by 4
  /// </summary>
  public sealed class Mod4Binary
  {
    private readonly  Byte[] data;
    public Mod4Binary(Byte[] data) 
    { 
      // Mod4Binary encapsulates the following domain rules
      if ((data == null) || (data.Length % 4 != 0))
        { throw new ArgumentException("Not evenly divisble by 4", "data"); }
      this.data = data; 
    }

    public Byte[] Data { get { return data; } }

    public override Boolean Equals (object obj)
    {
      var that = obj as Mod4Binary;
      if (that == null) return false;

      return this.data.SequenceEqual(that.data);
    }

    public override Int32 GetHashCode ()
    {
      const Int32 seed = -1640531527;
      return seed ^ this.data.GetHashCode();
    }

    public override String ToString ()
    {
      var buffer = new StringBuilder("Mod4Binary ");
      buffer.Append("[| ");
      this.data.Aggregate(buffer, (a,d) => a.AppendFormat("{0:X2}; ",d));
      buffer.Append(" |]");      
      return buffer.ToString();
    }
  }

  public static class Generators 
  {
    /// <summary>
    /// Allows FsCheck to randomly generate fszmq.Message instances
    /// </summary>
    public static Arbitrary<fszmq.Message> Message () 
    {
      // Generates an empty fszmq.Message
      var empty = Gen.Constant(new fszmq.Message());

      // Generates a fszmq.Message with random content
      var random = from data in Arb.Generate<NonNull<Byte[]>>() 
                   select new fszmq.Message(data.Get);
      
      // Creates a random distribution weighted in favor of messages with random content
      var distribution = Gen.Frequency(
          new WeightAndValue<GenMessage>(2, random)
         ,new WeightAndValue<GenMessage>(1, empty )
      );
      
      return Arb.From(distribution);
    }

    /// <summary>
    /// Allows FsCheck to randomly generate fszmq.Version instances
    /// </summary>
    public static Arbitrary<fszmq.Version> Version ()
    {
      // Generates a version when the actual values are unknown
      var unknown = Gen.Constant(fszmq.Version.Unknown);
      // Generates a version from random known values
      var random =  from data in Arb.Generate<NonNegativeInt>().Three()
                    let major     = data.Item1.Item
                    let minor     = data.Item2.Item
                    let revision  = data.Item3.Item
                    select fszmq.Version.NewVersion(major, minor, revision);
        
        
      // Creates a random distribution weighted in favor of version with random values
      var distribution = Gen.Frequency(
          new WeightAndValue<GenVersion>(2, random )
         ,new WeightAndValue<GenVersion>(1, unknown)
      );
      
      return Arb.From(distribution);
    }

    /// <summary>
    /// Allows FsCheck to randomly generate byte[] instances divisible by 4
    /// </summary>
    public static Arbitrary<Mod4Binary> Mod4Binary ()
    {
      // Mod4Binary encapsulates the following domain rules
      Func<Byte[],Boolean> isValid = (data) =>  (data != null)    && 
                                                (data.Length > 0) && 
                                                (data.Length % 4 == 0);
      
      return Arb.From(// generator
                      from data in Arb.Generate<Byte[]>()
                      where isValid(data)
                      select new Mod4Binary(data)
                      // shrinker
                     ,(input) => from data in Arb.Shrink(input.Data) 
                                 where  isValid(data)
                                 select new Mod4Binary(data));
    }
  }
}

File: fszmq.props/fszmq.props.vb/Generators.vb
Imports FsCheck.GenExtensions

Imports GenMessage = FsCheck.Gen(Of fszmq.Message)
Imports GenVersion = FsCheck.Gen(Of fszmq.Version)

Namespace fszmq.props

  ''' <summary>
  ''' A byte[] whose length is evenly divisible by 4
  ''' </summary>
  Public NotInheritable Class Mod4Binary
    Private ReadOnly data_() As Byte

    Public Sub New (ByVal data() As Byte)
      ' Mod4Binary encapsulates the following domain rules
      If data Is Nothing OrElse data.Length Mod 4 <> 0 Then 
        Throw New ArgumentException("Not evenly divisble by 4", "data")
      End If
      Me.data_ = data
    End Sub

    Public ReadOnly Property Data As Byte()
      Get
        Return Me.data_
      End Get
    End Property

    Public Overrides Function Equals(obj As Object) As Boolean
      Dim that = TryCast(obj, Mod4Binary)
      If that Is Nothing Then Return False

      Return Me.Data.SequenceEqual(that.Data)
    End Function

    Public Overrides Function GetHashCode() As Integer
      Const seed = -1640531527
      Return seed XOr Me.Data.GetHashCode()
    End Function
   
    Public Overrides Function ToString() As String
      Dim buffer As New StringBuilder("Mod4Binary ")
      buffer.Append("[| ")
      Me.Data.Aggregate(buffer, Function (a,d) a.AppendFormat("{0:X2}; ", d))
      buffer.Append(" |]")
      Return buffer.ToString()
    End Function

  End Class

  Module Generators

    ''' <summary>
    ''' Allows FsCheck to randomly generate fszmq.Message instances
    ''' </summary>
    Function Message () As Arbitrary(Of Message)
      ' Generates an empty fszmq.Message
      Dim empty = Gen.Constant(new Message())

      ' Generates a fszmq.Message with random content
      Dim random = From data In Arb.Generate(Of NonNull(Of Byte()))()
                   Select New Message(data.Get)
      
      ' Creates a random distribution weighted in favor of messages with random content
      Dim distribution = Gen.Frequency(
          new WeightAndValue(Of GenMessage)(2, random),
          new WeightAndValue(Of GenMessage)(1, empty )
      )
      
      Return Arb.From(distribution)
    End Function

    ''' <summary>
    ''' Allows FsCheck to randomly generate fszmq.Version instances
    ''' </summary>
    Function Version () As Arbitrary(Of Version)
      ' Generates a version when the actual values are unknown
      Dim unknown = Gen.Constant(Of Version)(Global.fszmq.Version.Unknown)

      ' Generates a version from random known values
      Dim random = From data In Arb.Generate(Of NonNegativeInt).Three()
                   Let major    = data.Item1.Item
                   Let minor    = data.Item2.Item
                   Let revision = data.Item3.Item
                   Select Global.fszmq.Version.NewVersion(major,minor,revision)

      ' Creates a random distribution weighted in favor of version with random values
      Dim distribution = Gen.Frequency(
          new WeightAndValue(Of GenVersion)(2, random),
          new WeightAndValue(Of GenVersion)(1, unknown)
      )

      Return Arb.From(distribution)
    End Function

    ''' <summary>
    ''' Allows FsCheck to randomly generate byte[] instances divisible by 4
    ''' </summary>
    Function Mod4Binary () As Arbitrary(Of Mod4Binary)
      ' Mod4Binary encapsulates the following domain rules
      Dim isValid = Function (ByVal data() As Byte) 
                      Return  (data IsNot Nothing)  AndAlso
                              (data.Length > 0)     AndAlso 
                              (data.Length Mod 4 = 0)
                    End Function

      Dim generator = From data In Arb.Generate(Of Byte())()
                      Where isValid(data)
                      Select New Mod4Binary(data)

      Dim shrinker =  Function (ByVal input As Mod4Binary)
                        Return  From data in Arb.Shrink(input.Data)
                                Where isValid(data)
                                Select New Mod4Binary(data)
                      End Function

      Return Arb.From(generator,shrinker)
    End Function
   
  End Module

End Namespace
